server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java

```diff
@@ -525,8 +535,10 @@ protected IndexItemResponse insert(Indexer indexer,
         ParsedDocument parsedDoc = rawIndexer == null ? indexer.index(item) : rawIndexer.index(item);
 
         // Replica must use the same values for undeterministic defaults/generated columns
-        if (indexer.hasUndeterministicSynthetics()) {
+        if (rawIndexer == null && indexer.hasUndeterministicSynthetics()) {
             item.insertValues(indexer.addGeneratedValues(item));
+        } else if (rawIndexer != null && rawIndexer.currentRowIndexer().hasUndeterministicSynthetics()) {
+            item.insertValues(rawIndexer.currentRowIndexer().addGeneratedValues(item));
```

Comment:
How about instead of `rawIndexer.currentRowIndexer().hasUndeterministicSynthetics()` we can `rawIndexer .hasUndeterministicSynthetics()` just to prevent exposing the internal `Indexer`.

Reply:
yep, sounds good